### PR TITLE
Implement a new controller to deploy a Service/Deploy from a Gateway

### PR DESCRIPTION
**Pull request by howardjohn**
This implements https://docs.google.com/document/d/1PfT7jdXSJ5iU_kU771YpV8Jq6l017Y65crOHJxXZORg/edit#

Open question: how do we want to enable it? I want it always-on, since using the API is very explicit, but that does mean all users will get elevated role (for svc/deployment write).

### CODE REVIEW

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
I donot think this is what istiod should do, i feels maybe this can goes to istio-operator

**Comment by howardjohn** [comment]
> I donot think this is what istiod should do, i feels maybe this can goes to istio-operator

IMO this must be in Istiod or not at all. The reason is this implements a core networking API (Gateway) and is primarily  to benefit the simpler use cases.

Requiring a core API to require a separate deployment goes against the Istiod design goals.

Requiring a separate deployment to do this defeats the simplicity goals of this design, as now they have to deploy a whole new component just to operate their gateways.


I think we have made a conscious decision with the Istiod movement to have a modularized monolith. All components (that reasonable can) should belong in Istiod; but they should be segmented and disable-able. This proposal meets that goal - if a user is not comfortable with this running in Istiod they are free to disable it, and I am happy to abstract it a bit more so it can run alone in a standalone `main.go` if users are interested in deploying just this controller as another deployment.

**Comment by hzxuzhonghu** [comment]
I would also want @istio/wg-product-security-maintainers @istio/wg-security-maintainers to have a look,

**Comment by howardjohn** [comment]
@hzxuzhonghu I have brought this up to PSWG/Security WG and got approval on the document.

**Comment by howardjohn** [comment]
@hzxuzhonghu I have also brought this up to TOC and there were not concerns

File: manifests/charts/istiod-remote/templates/clusterrole.yaml
Comment (costinm) [inline manifests/charts/istiod-remote/templates/clusterrole.yaml:40]: If possible, can you add a new ClusterRole object with the extra stuff, and a separate ClusterRoleBinding ? 
Seems better than having cluster roles that have different content based on flags.

Usually the cluster roles should be narrow and documented - the SA can have multiple bindings.

File: pilot/pkg/config/kube/crdclient/client.go
Comment (costinm) [inline pilot/pkg/config/kube/crdclient/client.go:118]: Some comment would be nice. Is this a new way to enable features only when the CRD is created ? Wouldn't a callback be better ( i.e. a map of CRD types to init functions ? )

File: pilot/pkg/config/kube/crdclient/client.go
Comment (costinm) [inline pilot/pkg/config/kube/crdclient/client.go:514]: I like the idea of starting various controllers if/when the CRD is detected, we should generalize it.

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (costinm) [inline pilot/pkg/config/kube/gateway/conversion.go:1120]: Comments. Why would it be managed if addresses are empty ?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go
Comment (costinm) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:219]: Seems like a pretty generic function. 
- don't we have it already ?
- should we move to /pkg ?

File: pilot/pkg/config/kube/gateway/templates.go
Comment (costinm) [inline pilot/pkg/config/kube/gateway/templates.go:32]: 'strdict' ? At least a comment explaining what it does. 

Does helm have a similar function ? It wouldn't be bad to reuse the names.

File: pilot/pkg/config/kube/gateway/templates/deployment.yaml
Comment (costinm) [inline pilot/pkg/config/kube/gateway/templates/deployment.yaml:15]: Would be really nice to use the same template elements/names as helm, it is harder to read/understand otherwise. And if this gets more complicated - we may reuse the helm-applying code we use in istioctl.

File: pilot/pkg/config/kube/gateway/templates/service-export.yaml
Comment (costinm) [inline pilot/pkg/config/kube/gateway/templates/service-export.yaml:2]: Why is ServiceExport here ?

Comment (costinm) [reply pilot/pkg/config/kube/gateway/templates/service-export.yaml:2]: I mean: we should not export it automatically AFAIK.

File: pkg/kube/client.go
Comment (costinm) [inline pkg/kube/client.go:404]: Do we need this ? Seems strange to not have needed it so far.

**Review by costinm — COMMENTED** [review]
Generally looks good, left a couple of comments but must can be addressed by adding TODOs and extra comments in the code, nothing fundamentally wrong.

**Comment by kyessenov** [comment]
Services are not intended as an automation API. There are quadratic costs associated, and vendors put a hard cap on number of services permitted. Is there an analysis of how many gateways are normally deployed?

**Comment by howardjohn** [comment]
> Services are not intended as an automation API.

Why not? There seems to be a large number of controllers existing that create Service, and all current implementations of gateway-api are planning the same.

> Is there an analysis of how many gateways are normally deployed?

Good question. We cannot get hard data since no one is using the API in prod but it would be reasonable to assume its less than Istio Gateways given our deployment mode is now 1:1 with Envoy whereas with Istio gateway its N:1.

For Istio gateways, we see:
p50 = 1 gateways
p95 = 4 gateways
p99 = 25 gateways

I do not expect problems here. Users will be well aware that a Gateway resource has potential underlying cost, as ALL implementations of gateway-api will have this behavior.

File: pilot/pkg/config/kube/gateway/templates/service-export.yaml
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/templates/service-export.yaml:2]: Its not automatic, but allow users a way to export from the Gateway, vs needing to create a ServiceExport on their own (which somewhat leaks the fact there is a Service created).

Thinking about it more its probably not very useful

File: pilot/pkg/config/kube/crdclient/client.go
Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:118]: Yes, added some comments.

A callback seems nice but I think it may add some complications around the startup logic that makes it a bit more complex. Right now the startup handles releasing leader lock on shutdown, etc that would be a bit harder with callback. If there are more use cases  that callback would be better we can always add it later I think?

File: pilot/pkg/config/kube/gateway/conversion.go
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1120]: Added comment in code, copying here:
```go
// isManaged checks if a Gateway is managed (ie we create the Deployment and Service) or unmanaged.
// This is based on the address field of the spec. If address is set, it should point to an existing
// Service that handles the gateway traffic. If it is not set, we will consider it managed and provision the Service.
// While there is no defined standard for this in the API yet, it is tracked in https://github.com/kubernetes-sigs/gateway-api/issues/892.
// So far, this mirrors how out of clusters work (address set means to use existing IP, unset means to provision one),
// and there has been growing consensus on this model for in cluster deployments.
func isManaged(gw *k8s.GatewaySpec) bool {
	return len(gw.Addresses) == 0
}
```

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:219]: We don't. The only thing really apply templates to cluster is the operator code, which use controller-runtime which is incompatible with our needs for Istiod (discussed at length in the DeploymentController comments).

There is the new `controllers` package so moving it there seems reasonable

File: pilot/pkg/config/kube/gateway/templates.go
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/templates.go:32]: Yeah the problem is it returns the wrong type unfortunately. added a comment

File: pkg/kube/client.go
Comment (howardjohn) [reply pkg/kube/client.go:404]: no, was for debugging

**Review by costinm — APPROVED** [review]
On creating Service and its cost: KNative and others do the same thing, and AFAIK the consensus is that subsets will also be represented by Service ( like KNative does).

We may start using IPV6, and k8s may need to optimize how they implement Service - but in case of Gateway it is really not a problem since we would create the same service if helm was used.

**Comment by howardjohn** [comment]
@hzxuzhonghu ptal.

**Review by hzxuzhonghu — COMMENTED** [review]
The implement looks good to me, I am concerned about the number of gateways as well

**Comment by howardjohn** [comment]
> The implement looks good to me, I am concerned about the number of gateways as well

Just like today, the user can choose how many they want. If they only need 1 then they should just create 1 gateway.

**Review by hzxuzhonghu — APPROVED** [review]
approve since it is optional

**Pull request merged**
